### PR TITLE
Optimizer rules that combine shifts and masks.

### LINKED ISSUE #6214: Optimizer: Weird shifts still unoptimized
The following sequence is part of bytecode generated from Solidity after the optimizer has run.

`and(0xff, shr(0xf8, shl(0xf8, shr(0xf8, x))))`

It can obviously be simplified to `shr(0xf8, x)`.

Related: https://github.com/ethereum/solidity/issues/5810

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/6214

### CODE REVIEW

File: libevmasm/RuleList.h [hunk 1]
```diff
@@ -47,6 +47,11 @@ template <class S> S modWorkaround(S const& _a, S const& _b)
 	return (S)(bigint(_a) % bigint(_b));
 }
 
+inline u256 shlWorkaround(u256 const& _x, unsigned _amount)
```
Comment (axic) [inline libevmasm/RuleList.h:52]: There is a helper for this in libdevcore, IIRC.

**Comment by chriseth** [comment]
I'll add some more tests.

**Comment by chriseth** [comment]
Added more tests.

File: libevmasm/RuleList.h (hunk 1 above)
Comment (axic) [reply libevmasm/RuleList.h:52]: See https://github.com/ethereum/solidity/blob/develop/libdevcore/CommonData.h#L191

File: libdevcore/CommonData.h [hunk 2]
```diff
@@ -188,12 +188,6 @@ inline bytes toCompactBigEndian(uint8_t _val, unsigned _min = 0)
 	return (_min || _val) ? bytes{ _val } : bytes{};
 }
 
-/// Workarounds shift left bug in boost <1.65.1.
```
Comment (axic) [inline libdevcore/CommonData.h:191]: Why move this and not keep it here?

Comment (chriseth) [reply libdevcore/CommonData.h:191]: It is not used anywhere else and all the other workarounds are also in the rules.

**Review by leonardoalt — DISMISSED** [review]

**Comment by bshastry** [comment]
I've not been able to extensively fuzz this PR due to other blockers but this PR does pass against the current fuzzing corpus for the yul optimizer.

**Comment by chriseth** [comment]
Does the corpus contain enough shift instructions? We should fix the blockers....

File: libdevcore/CommonData.h (hunk 2 above)
Comment (axic) [reply libdevcore/CommonData.h:191]: How? There were rules in RuleList which used it?

Comment (chriseth) [reply libdevcore/CommonData.h:191]: Sure, I changed them to use the implementation inside RuleList.

**Comment by bshastry** [comment]
> Does the corpus contain enough shift instructions? We should fix the blockers....

Yes, it should. To be sure, I will monitor the coverage from the fuzzing runs to see if the newly added lines in this PR have been hit.

**Comment by bshastry** [comment]
I checked coverage of the newly added rules to the RuleList and it looks like the fuzzer hits them consistently (several hundred times with different inputs) without finding reporting errors.

File: libevmasm/RuleList.h [hunk 3]
```diff
@@ -47,6 +47,13 @@ template <class S> S modWorkaround(S const& _a, S const& _b)
 	return (S)(bigint(_a) % bigint(_b));
 }
 
+// This works around a bug fixed with Boost 1.64.
+// https://www.boost.org/doc/libs/1_68_0/libs/multiprecision/doc/html/boost_multiprecision/map/hist.html#boost_multiprecision.map.hist.multiprecision_2_3_1_boost_1_64
+inline u256 shlWorkaround(u256 const& _x, unsigned _amount)
+{
+	return u256((bigint(_x) << _amount) & u256(-1));
```
Comment (ekpyron) [inline libevmasm/RuleList.h:54]: Is the ``& u256(-1)`` really required here? That should be implicit in the conversion to ``u256`` anyways... Seems to be a pretty weird boost bug then...

**Review by ekpyron — COMMENTED** [review]
I'm pretty sure the rules are correct - I just looked into formally verifying them with Isabelle/HOL and realized that doing so would take quite some time, though.

File: libevmasm/RuleList.h (hunk 3 above)
Comment (chriseth) [reply libevmasm/RuleList.h:54]: It is required for downcasting from u256, but I'm not sure for bigint. The multiprecision documentation claims that downcasting an out-of-range value does not result in bits being clipped but in the largest value that is in range. It also claims that the implementation had a bug previously.

**Comment by leonardoalt** [comment]
`shlWorkaround` and the first two rules in this PR are checked in https://github.com/leonardoalt/solidity_proofs/pull/1.
The size of the variables is scaled down to 64 bits, where we assume that the result is the same for 256.

**Comment by leonardoalt** [comment]
The last 4 rules are now also proved in https://github.com/leonardoalt/solidity_proofs/pull/1 for 128 bit values.

**Pull request merged**